### PR TITLE
Ignore metadata of deleted indices at start

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today in 6.x it is possible to add an index tombstone to the graveyard without
deleting the corresponding index metadata, because the deletion is slightly
deferred. If you shut down the node and upgrade to 7.x when in this state then
the node will fail to apply any cluster states, reporting

    java.lang.IllegalStateException: Cannot delete index [...], it is still part of the cluster state.

This commit addresses this situation by skipping over any index metadata with a
corresponding tombstone, allowing this metadata to be cleaned up by the 7.x
node.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Relates https://discuss.elastic.co/t/upgrade-from-6-8-3-to-7-x-x-results-in-failed-to-apply-updated-cluster-state/207111/3

**Comment by DaveCTurner** [comment]
This needs more work because it doesn't address the case where you're in this state in a rolling upgrade.

**Comment by DaveCTurner** [comment]
> This needs more work because it doesn't address the case where you're in this state in a rolling upgrade.

Discussed this and decided it's ok for a rolling upgrade to fall back on a full cluster restart if it happens to be in this state, and this PR will allow the full cluster restart to proceed.

File: server/src/test/java/org/elasticsearch/gateway/GatewayIndexStateIT.java [hunk 1]
```diff
@@ -508,6 +515,42 @@ public void testArchiveBrokenClusterSettings() throws Exception {
...
+        internalCluster().startNode();
+        createIndex("test", Settings.builder()
+            .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0)
+            .build());
+        ensureGreen("test");
+
+        final MetaData metaData = internalCluster().getInstance(ClusterService.class).state().metaData();
+        final Path[] paths = internalCluster().getInstance(NodeEnvironment.class).nodeDataPaths();
+        writeBrokenMeta(metaStateService -> {
```
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/gateway/GatewayIndexStateIT.java]: I know this is not your change, but the name "writeBrokenMeta" looks invalid for two reasons - in this test we write well-formed metadata, this method performs full-cluster restart and I think this should be reflected in the method name.

**Review by andrershov — COMMENTED** [review]
Looks good, I've left one request to rename the function name

File: server/src/test/java/org/elasticsearch/gateway/GatewayIndexStateIT.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/gateway/GatewayIndexStateIT.java]: A more substantial change to this method (including fixing its name) is incoming in https://github.com/elastic/elasticsearch/pull/48733/files#diff-a53ee618ca95b1bde55d7f5508a03d6aR511.

Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/gateway/GatewayIndexStateIT.java]: However the metadata written here is well-formed but still broken - it contains a tombstone for an index that is not properly deleted.

**Review by andrershov — APPROVED** [review]
LGTM

**Comment by dpeddi** [comment]
I got this error while upgrading from 6.7 to 7.0.1.
After reading this I tried another upgrade upgrade to 7.10.2 but the error is still present.

Just for other users/readers, I was able to restart my cluster by manually removing all the state with the latest lucene/luke after reading the "Cannot delete index [...], it is still part of the cluster state." message for each node start

**Comment by DaveCTurner** [comment]
> Just for other users/readers, I was able to restart my cluster by manually removing all the state with the latest lucene/luke

For the sake of other users/readers, this advice is extremely dangerous and we do not recommend following it. Editing the contents of the data path using a tool like Luke can result in arbitrary and silent data loss.

@dpeddi the actual solution is not to upgrade to 7.0.1, this version is already long past [EOL](https://www.elastic.co/support/eol). The [current recommendation](https://www.elastic.co/guide/en/elasticsearch/reference/current/setup-upgrade.html) is to upgrade to the latest 6.8 and then the latest 7.x (7.10.2 at time of writing).

**Pull request merged**